src/udp_engine.cpp

```diff
@@ -145,11 +149,29 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
 #endif
 
         if (address->resolved.udp_addr->is_mcast ()) {
-            struct ip_mreq mreq;
-            mreq.imr_multiaddr = address->resolved.udp_addr->multicast_ip ();
-            mreq.imr_interface = address->resolved.udp_addr->interface_ip ();
-            rc = setsockopt (fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char *) &mreq,
-                             sizeof (mreq));
+            const ip_addr_t *mcast_addr =
+              address->resolved.udp_addr->target_addr ();
+
+            if (mcast_addr->family () == AF_INET) {
+                struct ip_mreq mreq;
+                mreq.imr_multiaddr = mcast_addr->ipv4.sin_addr;
+                mreq.imr_interface = bind_addr->ipv4.sin_addr;
+
+                rc = setsockopt (fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
+                                 (char *) &mreq, sizeof (mreq));
+            } else if (mcast_addr->family () == AF_INET6) {
+                struct ipv6_mreq mreq;
+
+                mreq.ipv6mr_multiaddr = mcast_addr->ipv6.sin6_addr;
+                mreq.ipv6mr_interface = 0; // XXX fixme?
```

Comment:
So here I'm having a bit of a conundrum: unlike the IPv4 code `struct ipv6_mreq` doesn't let you specify the IP address of the interface to bind but rather the interface number. It's trivial to lookup if we have the interface name using `if_nametoindex` but of course at this point we don't have the interface name anymore and on top of that ZMQ URLs lets you specify an interface by IP: `udp://[abcd::1];[ff00::1]:5555`.

The easy solution would be to only support specifying a source interface by giving the interface name but that's obviously limiting and not very consistent. The hard solution would be to have a method to look up an interface index based on an IP (pretty much the opposite of `zmq::ip_resolver_t::resolve_nic_name`). Of course the tedious part is that this code is very platform-specific.

I think the latter solution is the way to go but I'd like to have some feedback on that.

Reply:
If you want to start by only supporting the interface name, and then later on add support for the IP, that's fine given this socket type is a work in progress